{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/polinliv/livanou/blob/main/%CE%91%CE%BD%CF%84%CE%AF%CE%B3%CF%81%CE%B1%CF%86%CE%BF_greedy_motif_search_demo.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**GreedyMotifSearch**\n",
        "\n",
        "Based on code from:\n",
        "1. Compeau & Pevzner, Bioinformatics Algorithms\n",
        "2. https://github.com/rivers-yao/Coursework/blob/master/Bioinformatics%20Algorithm/Week2"
      ],
      "metadata": {
        "id": "Ot4Kc5Cr8uGI"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import math"
      ],
      "metadata": {
        "id": "0-IagHnK8r0o"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "JpOrmVAR8hgd"
      },
      "outputs": [],
      "source": [
        "# Auxialiary Functions\n",
        "def SymbolToNumber(Symbol):\n",
        "    \"\"\"\n",
        "    Converts base to number (in lexicograpical order)\n",
        "\n",
        "    Symbol: the letter to be converted (str)\n",
        "\n",
        "    Returns: the number corresponding to that base (int)\n",
        "    \"\"\"\n",
        "    if Symbol == \"A\":\n",
        "        return 0\n",
        "    elif Symbol == \"C\":\n",
        "        return 1\n",
        "    elif Symbol == \"G\":\n",
        "        return 2\n",
        "    elif Symbol == \"T\":\n",
        "        return 3\n",
        "    else:\n",
        "        print(f'Unexpected symbol {Symbol} in function SymbolToNumber()')\n",
        "        return None\n",
        "\n",
        "\n",
        "def NumberToSymbol(index):\n",
        "    \"\"\"\n",
        "    Finds base from number (in lexicographical order)\n",
        "\n",
        "    index: the number to be converted (int)\n",
        "\n",
        "    Returns: the base corresponding to index (str)\n",
        "    \"\"\"\n",
        "    if index == 0:\n",
        "        return str(\"A\")\n",
        "    elif index == 1:\n",
        "        return str(\"C\")\n",
        "    elif index == 2:\n",
        "        return str(\"G\")\n",
        "    elif index == 3:\n",
        "        return str(\"T\")\n",
        "\n",
        "\n",
        "def HammingDistance(p, q):\n",
        "    \"\"\"\n",
        "    Finds the number of mismatches between 2 DNA segments of equal lengths\n",
        "\n",
        "    p: first DNA segment (str)\n",
        "\n",
        "    q: second DNA segment (str)\n",
        "\n",
        "    Returns: number of mismatches (int)\n",
        "    \"\"\"\n",
        "    return sum(s1 != s2 for s1, s2 in zip(p, q))\n",
        "\n",
        "\n",
        "def window(s, k):\n",
        "    for i in range(1 + len(s) - k):\n",
        "        yield s[i:i+k]\n",
        "\n",
        "\n",
        "def ProfileMostProbable(Text, k, Profile):\n",
        "    \"\"\"\n",
        "    Finds a k-mer that was most likely to be generated by profile among\n",
        "    all k-mers in Text\n",
        "\n",
        "    Text: given DNA segment (str)\n",
        "\n",
        "    k: length of pattern (int)\n",
        "\n",
        "    Profile: a 4x4 matrix (list)\n",
        "\n",
        "    Returns: profile-most probable k-mer (str)\n",
        "    \"\"\"\n",
        "    letter = [[] for key in range(k)]\n",
        "    probable = \"\"\n",
        "    hamdict = {}\n",
        "    index = 1\n",
        "    for a in range(k):\n",
        "        for j in \"ACGT\":\n",
        "            letter[a].append(Profile[j][a])\n",
        "    for b in range(len(letter)):\n",
        "        number = max(letter[b])\n",
        "        probable += str(NumberToSymbol(letter[b].index(number)))\n",
        "    for c in window(Text, k):\n",
        "        for x in range(len(c)):\n",
        "            y = SymbolToNumber(c[x])\n",
        "            index *= float(letter[x][y])\n",
        "        hamdict[c] = index\n",
        "        index = 1\n",
        "    for pat, ham in hamdict.items():\n",
        "        if ham == max(hamdict.values()):\n",
        "            final = pat\n",
        "            break\n",
        "    return final\n",
        "\n",
        "\n",
        "def Count(Motifs):\n",
        "    \"\"\"\n",
        "    Documentation here\n",
        "    \"\"\"\n",
        "    count = {}\n",
        "    k = len(Motifs[0])\n",
        "    for symbol in \"ACGT\":\n",
        "        count[symbol] = []\n",
        "        for i in range(k):\n",
        "            count[symbol].append(0)\n",
        "    t = len(Motifs)\n",
        "    for i in range(t):\n",
        "        for j in range(k):\n",
        "            symbol = Motifs[i][j]\n",
        "            count[symbol][j] += 1\n",
        "    return count\n",
        "\n",
        "\n",
        "def FindConsensus(motifs):\n",
        "    \"\"\"\n",
        "    Finds a consensus sequence for given list of motifs\n",
        "\n",
        "    motifs: a list of motif sequences (list)\n",
        "\n",
        "    Returns: consensus sequence of motifs (str)\n",
        "    \"\"\"\n",
        "    consensus = \"\"\n",
        "    for i in range(len(motifs[0])):\n",
        "        countA, countC, countG, countT = 0, 0, 0, 0\n",
        "        for motif in motifs:\n",
        "            if motif[i] == \"A\":\n",
        "                countA += 1\n",
        "            elif motif[i] == \"C\":\n",
        "                countC += 1\n",
        "            elif motif[i] == \"G\":\n",
        "                countG += 1\n",
        "            elif motif[i] == \"T\":\n",
        "                countT += 1\n",
        "        if countA >= max(countC, countG, countT):\n",
        "            consensus += \"A\"\n",
        "        elif countC >= max(countA, countG, countT):\n",
        "            consensus += \"C\"\n",
        "        elif countG >= max(countC, countA, countT):\n",
        "            consensus += \"G\"\n",
        "        elif countT >= max(countC, countG, countA):\n",
        "            consensus += \"T\"\n",
        "    return consensus\n",
        "\n",
        "\n",
        "def ProfileMatrix(motifs):\n",
        "    \"\"\"\n",
        "    Finds the profile matrix for given list of motifs\n",
        "\n",
        "    motifs: list of motif sequences (list)\n",
        "\n",
        "    Returns: the profile matrix for motifs (list)\n",
        "    \"\"\"\n",
        "    Profile = {}\n",
        "    A, C, G, T = [], [], [], []\n",
        "    for j in range(len(motifs[0])):\n",
        "        countA, countC, countG, countT = 0, 0, 0, 0\n",
        "        for motif in motifs:\n",
        "            if motif[j] == \"A\":\n",
        "                countA += 1\n",
        "            elif motif[j] == \"C\":\n",
        "                countC += 1\n",
        "            elif motif[j] == \"G\":\n",
        "                countG += 1\n",
        "            elif motif[j] == \"T\":\n",
        "                countT += 1\n",
        "        A.append(countA)\n",
        "        C.append(countC)\n",
        "        G.append(countG)\n",
        "        T.append(countT)\n",
        "    Profile[\"A\"] = A\n",
        "    Profile[\"C\"] = C\n",
        "    Profile[\"G\"] = G\n",
        "    Profile[\"T\"] = T\n",
        "    return Profile\n",
        "\n",
        "\n",
        "def Score(motifs):\n",
        "    \"\"\"\n",
        "    Finds score of motifs relative to the consensus sequence\n",
        "\n",
        "    motifs: a list of given motifs (list)\n",
        "\n",
        "    Returns: score of given motifs (int)\n",
        "    \"\"\"\n",
        "    consensus = FindConsensus(motifs)\n",
        "    score = 0.0000\n",
        "    for motif in motifs:\n",
        "        score += HammingDistance(consensus, motif)\n",
        "    #print(score)\n",
        "    return round(score, 4)"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Greedy Motif Search Algorithm**"
      ],
      "metadata": {
        "id": "OmsvUCG9jfrT"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def GreedyMotifSearch(DNA, k):\n",
        "    print(f'Greedy search for motifs of length {k}. ')\n",
        "    print(f'The best solution that is found is returned.')\n",
        "\n",
        "    t = len(DNA)\n",
        "    bestMotifs = []\n",
        "    bestScore = math.inf\n",
        "    for string in DNA:\n",
        "        bestMotifs.append(string[:k])\n",
        "    base = DNA[0]\n",
        "    for i in window(base, k):\n",
        "        # Change here. Should start with one element in motifs and build up.\n",
        "        # As in the line \"motifs ‚Üê list with only Dna[0](i,k)\"\n",
        "        # newMotifs = []\n",
        "        newMotifs = [i]\n",
        "        # Change here to iterate over len(DNA).\n",
        "        # Should go through \"for j from 1 to |Dna| - 1\"\n",
        "        # for j in range(t):\n",
        "        for j in range(1, len(DNA)):\n",
        "            # Change here. Should build up motifs and build profile using them.\n",
        "            # profile = ProfileMatrix([i])\n",
        "            profile = ProfileMatrix(newMotifs)\n",
        "            probable = ProfileMostProbable(DNA[j], k, profile)\n",
        "            newMotifs.append(probable)\n",
        "\n",
        "        # Change to < rather < = to ensure getting the most recent hit. As referenced in the instructions:\n",
        "        # If at any step you find more than one Profile-most probable k-mer in a given string, use the one occurring **first**.\n",
        "        new_score = Score(newMotifs)\n",
        "        if new_score < bestScore:\n",
        "        #if Score(newMotifs) <= bestScore:\n",
        "            bestScore = new_score\n",
        "            bestMotifs = newMotifs\n",
        "    return bestMotifs"
      ],
      "metadata": {
        "id": "OzUsHsmU9T-A"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Data: DNA Sequences**"
      ],
      "metadata": {
        "id": "C5tjIaSDjVVk"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "DNA_sequences = ['GTATCTACTCACGTTGCTACCTGCGGTTCCGATTAGGATCAGGTTATAACGTAGTCGCCCTGATATGACTACATCGTTGCAGGCTTGTTACACGTACTAACATATTGGAAAACG',\n",
        "'TGCCCACTCCCTGATACGCGTTGACCGATCGTGGTGATTTCTTGATCTAAGTCGCCACCCTCTTCAAGGGCTATACATCAACAGAAATTTGTCTGTTTTAGGTCTGTGCCCGCC',\n",
        "'CTAGACGTAGCGATATGTCGTATGGGACGGGATGTCGCGTGCAAAAATGACCCCTCGGAACCGGCTAAGGCAGCTGTCTATGAGTGGTATTCCTGCTAGGCGTTAACTCGGACT',\n",
        "'GAGGACCCGGGTACTTGCAGCAAGGCGTGAAAGAAGAAATAATGGTCTATAAAGTGCCCCTGATTCCTGCTCTCAAGGCACCTCGGTGGGGTGCGTTAGGCATAATTTGTTTGT',\n",
        "'GCCAACACATCCGACGAGAGGTATCAAGCGTACCGCACGTTCACAGGAGATTCCATGGCGACTTAAGTAGATCGTCGTGCACAAAGGCGCCAGACCGATGATACGCGTACAATC',\n",
        "'ATTCGTCGACGACACCTCCTGCTCATAGTCGCGACCTTAACCTTCTCCACCTTGACTACCCCGGCCTGAGCAAACGCTTCATTAAACATAGAAATTAACCAAGACGGTGGATAG',\n",
        "'CGCGGCTCTAACTCAAACCCGACAACTCCGAGGATCCCGGGTGAGCATTATCGTGGGCAATTTTACTGCCGGGTTCTGGTAGTAGTCGCGCAGGTAACGCGCGAAGCTTGTGTC',\n",
        "'TCAATGACAACGGGGCAGCTGGTCGTGGACGCCCCGGTGCGCCCCCGTAAACTACTAGGTAAATCTCACTACGCTCGGGCAACGGAAAACTCAACCAGAACGGACGAGCAAATC',\n",
        "'TTCAGCAGACATGAGTACTATTTAGCTGTCTACAGAAGCTGGTCCTCCACGGACTGTCTGATTGCTAGGGGCCCAAATGATTACTTATATCGTACACATATTGCAGATTTGAAA',\n",
        "'CGCACCGCATAGTCACAGACGGGCCTTGTCGTCCTCGGGCAGGAGCGTATCTCGACAGTGGCCGCTACCATCGTACATGGTCCAGAGTAGAATAGGAATCACTCGTGTGAACAT',]\n"
      ],
      "metadata": {
        "id": "dzsZ7G1i9V4f"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print(DNA_sequences)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "VqtYA4Y8H8Yx",
        "outputId": "d32474a3-82f9-49ef-f97a-3a8813e128de"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "['GTATCTACTCACGTTGCTACCTGCGGTTCCGATTAGGATCAGGTTATAACGTAGTCGCCCTGATATGACTACATCGTTGCAGGCTTGTTACACGTACTAACATATTGGAAAACG', 'TGCCCACTCCCTGATACGCGTTGACCGATCGTGGTGATTTCTTGATCTAAGTCGCCACCCTCTTCAAGGGCTATACATCAACAGAAATTTGTCTGTTTTAGGTCTGTGCCCGCC', 'CTAGACGTAGCGATATGTCGTATGGGACGGGATGTCGCGTGCAAAAATGACCCCTCGGAACCGGCTAAGGCAGCTGTCTATGAGTGGTATTCCTGCTAGGCGTTAACTCGGACT', 'GAGGACCCGGGTACTTGCAGCAAGGCGTGAAAGAAGAAATAATGGTCTATAAAGTGCCCCTGATTCCTGCTCTCAAGGCACCTCGGTGGGGTGCGTTAGGCATAATTTGTTTGT', 'GCCAACACATCCGACGAGAGGTATCAAGCGTACCGCACGTTCACAGGAGATTCCATGGCGACTTAAGTAGATCGTCGTGCACAAAGGCGCCAGACCGATGATACGCGTACAATC', 'ATTCGTCGACGACACCTCCTGCTCATAGTCGCGACCTTAACCTTCTCCACCTTGACTACCCCGGCCTGAGCAAACGCTTCATTAAACATAGAAATTAACCAAGACGGTGGATAG', 'CGCGGCTCTAACTCAAACCCGACAACTCCGAGGATCCCGGGTGAGCATTATCGTGGGCAATTTTACTGCCGGGTTCTGGTAGTAGTCGCGCAGGTAACGCGCGAAGCTTGTGTC', 'TCAATGACAACGGGGCAGCTGGTCGTGGACGCCCCGGTGCGCCCCCGTAAACTACTAGGTAAATCTCACTACGCTCGGGCAACGGAAAACTCAACCAGAACGGACGAGCAAATC', 'TTCAGCAGACATGAGTACTATTTAGCTGTCTACAGAAGCTGGTCCTCCACGGACTGTCTGATTGCTAGGGGCCCAAATGATTACTTATATCGTACACATATTGCAGATTTGAAA', 'CGCACCGCATAGTCACAGACGGGCCTTGTCGTCCTCGGGCAGGAGCGTATCTCGACAGTGGCCGCTACCATCGTACATGGTCCAGAGTAGAATAGGAATCACTCGTGTGAACAT']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Greedy Motif Search: Computation**"
      ],
      "metadata": {
        "id": "1bOAlxHBHaAs"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "k = 15\n",
        "best_motifs = GreedyMotifSearch(DNA_sequences, k)\n",
        "\n",
        "print('The best motifs found:')\n",
        "for i, s in enumerate(DNA_sequences):\n",
        "    print(f'{str(i).rjust(2)}, motif location: {DNA_sequences[i].find(best_motifs[i])}, motif: {best_motifs[i]}')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "h4F8DXCjGdIp",
        "outputId": "6581569b-9744-42ef-9a33-a460e29632cf"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Greedy search for motifs of length 15. \n",
            "The best solution that is found is returned.\n",
            "The best motifs found:\n",
            " 0, motif location: 83, motif: CTTGTTACACGTACT\n",
            " 1, motif location: 0, motif: TGCCCACTCCCTGAT\n",
            " 2, motif location: 0, motif: CTAGACGTAGCGATA\n",
            " 3, motif location: 0, motif: GAGGACCCGGGTACT\n",
            " 4, motif location: 96, motif: GATGATACGCGTACA\n",
            " 5, motif location: 0, motif: ATTCGTCGACGACAC\n",
            " 6, motif location: 83, motif: AGTCGCGCAGGTAAC\n",
            " 7, motif location: 36, motif: GTGCGCCCCCGTAAA\n",
            " 8, motif location: 40, motif: GGTCCTCCACGGACT\n",
            " 9, motif location: 10, motif: AGTCACAGACGGGCC\n"
          ]
        }
      ]
    }
  ]
}